### PR TITLE
Motor subsystem power -> core

### PR DESCRIPTION
**Pull request by Nitay4**

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/motorSubsystem/MotorSubsystem.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+	public MotorSubsystem withVoltageControl(IRequest<Double> voltageRequest, InputSignal<Double> voltageSignal) {
+		this.voltageRequest = voltageRequest;
+		this.voltageSignal = voltageSignal;
+		return this;
+	}
+
+	public Double getAppliedVoltage() {
+		if (voltageSignal == null) {
+			throw new NullPointerException("the voltage signal is null, try using: '.withVoltageControl'");
```
Comment (Yotam7) [inline src/main/java/frc/robot/subsystems/motorSubsystem/MotorSubsystem.java:30]:
```suggestion
			throw new NullPointerException("The voltage signal is null. try using '.withVoltageControl'");
```

File: src/main/java/frc/robot/subsystems/motorSubsystem/MotorSubsystem.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+		motor.stop();
+	}
+
+	public MotorCommandBuilder getCommandBuilder() {
+		return commandBuilder;
+	}
+
+	public void setVoltage(double voltage) {
+		if (voltageRequest == null) {
+			throw new NullPointerException("the voltage request is null, try using: '.withVoltageControl'");
```
Comment (Yotam7) [inline src/main/java/frc/robot/subsystems/motorSubsystem/MotorSubsystem.java:45]:
```suggestion
			throw new NullPointerException("The voltage request is null. try using '.withVoltageControl'");
```

File: src/main/java/frc/robot/subsystems/motorSubsystem/MotorSubsystem.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	public void updateInputs() {
+		motor.updateInputs(voltageSignal);
+	}
+
+	public void applyRequests() {
+		if (lastChangedRequest != null) {
+			motor.applyRequest(lastChangedRequest);
+		}
+	}
+	@Override
```
Comment (Yotam7) [inline src/main/java/frc/robot/subsystems/motorSubsystem/MotorSubsystem.java:65]:
```suggestion
	}

	@Override
```

File: src/main/java/frc/robot/subsystems/motorSubsystem/MotorSubsystem.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+	protected void subsystemPeriodic() {
+		updateInputs();
+		applyRequests();
+	}
```
Comment (Sh-Pe) [inline src/main/java/frc/robot/subsystems/motorSubsystem/MotorSubsystem.java:69]:
```suggestion
	}
	
```

File: src/main/java/frc/robot/subsystems/motorSubsystem/MotorCommandBuilder.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+	public MotorCommandBuilder(MotorSubsystem motor) {
+		this.motor = motor;
+	}
+
+	public Command setPower(DoubleSupplier powerSupplier) {
+		return new ExecuteEndCommand(() -> motor.setPower(powerSupplier.getAsDouble()), () -> {}, motor);
+	}
+
+	public Command setPower(double power) {
```
Comment (asaf8888) [inline src/main/java/frc/robot/subsystems/motorSubsystem/MotorCommandBuilder.java:21]: RunCommand executes the Runnable in loop this and the function below can use an InstantCommand
the function above can use RunCommand

**Review by bossyshrimp68 — DISMISSED** [review]

File: src/main/java/frc/robot/subsystems/motorSubsystem/MotorCommandBuilder.java (hunk 5 above)
Comment (asaf8888) [reply src/main/java/frc/robot/subsystems/motorSubsystem/MotorCommandBuilder.java:21]: also all three commands should require the subsystem

File: src/main/java/frc/robot/subsystems/motorSubsystem/MotorCommandBuilder.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+import edu.wpi.first.wpilibj2.command.Command;
+import edu.wpi.first.wpilibj2.command.RunCommand;
+import frc.utils.utilcommands.ExecuteEndCommand;
+
+import java.util.function.DoubleSupplier;
+
+public class MotorCommandBuilder {
+
+	private MotorSubsystem motor;
```
Comment (asaf8888) [inline src/main/java/frc/robot/subsystems/motorSubsystem/MotorCommandBuilder.java:11]: perhaps motorSubsystem instead?

File: src/main/java/frc/robot/subsystems/motorSubsystem/MotorCommandBuilder.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+	public MotorCommandBuilder(MotorSubsystem motor) {
+		this.motor = motor;
+	}
+
+	public Command setPower(DoubleSupplier powerSupplier) {
```
Comment (asaf8888) [inline src/main/java/frc/robot/subsystems/motorSubsystem/MotorCommandBuilder.java:17]: shouldnt there be a set voltage command?

File: src/main/java/frc/robot/subsystems/motorSubsystem/MotorCommandBuilder.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+import java.util.function.DoubleSupplier;
+
+public class MotorCommandBuilder {
+
+	private MotorSubsystem motorSubsystem;
+
+	public MotorCommandBuilder(MotorSubsystem motorSubsystem) {
+		this.motorSubsystem = motorSubsystem;
+	}
+
```
Comment (YHerm) [inline src/main/java/frc/robot/subsystems/motorSubsystem/MotorCommandBuilder.java:15]: maybe add the setVoltageByDashboard too

File: src/main/java/frc/robot/subsystems/motorSubsystem/MotorSubsystem.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+public class MotorSubsystem extends GBSubsystem {
+
+	private final ControllableMotor motor;
+	private final MotorCommandBuilder commandBuilder;
+	private IRequest<Double> voltageRequest;
+	private InputSignal<Double> voltageSignal;
+	private IRequest lastChangedRequest;
+
+
```
Comment (YHerm) [inline src/main/java/frc/robot/subsystems/motorSubsystem/MotorSubsystem.java:16]: rm space

File: src/main/java/frc/robot/subsystems/motorSubsystem/MotorSubsystem.java [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+		this.commandBuilder = new MotorCommandBuilder(this);
+	}
+
+	public MotorSubsystem withVoltageControl(IRequest<Double> voltageRequest, InputSignal<Double> voltageSignal) {
+		this.voltageRequest = voltageRequest;
+		this.voltageSignal = voltageSignal;
+		return this;
+	}
+
+	public Double getAppliedVoltage() {
```
Comment (YHerm) [inline src/main/java/frc/robot/subsystems/motorSubsystem/MotorSubsystem.java:29]:
```suggestion
	public Double getVoltage() {
```
that how all names now

File: src/main/java/frc/robot/subsystems/motorSubsystem/MotorSubsystem.java [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+		return voltageSignal.getLatestValue();
+	}
+
+	public void stop() {
+		motor.stop();
+	}
+
+	public MotorCommandBuilder getCommandBuilder() {
+		return commandBuilder;
+	}
```
Comment (YHerm) [inline src/main/java/frc/robot/subsystems/motorSubsystem/MotorSubsystem.java:42]: put under constructor

File: src/main/java/frc/robot/subsystems/motorSubsystem/MotorSubsystem.java [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+	public Double getAppliedVoltage() {
+		if (voltageSignal == null) {
+			throw new NullPointerException("The voltage signal is null. try using '.withVoltageControl'");
+		}
+		return voltageSignal.getLatestValue();
+	}
+
+	public void stop() {
+		motor.stop();
+	}
```
Comment (YHerm) [inline src/main/java/frc/robot/subsystems/motorSubsystem/MotorSubsystem.java:38]: organize funcs in this class

**Review by YHerm — COMMENTED** [review]
please split the power to pr with brake🥹

File: src/main/java/frc/robot/subsystems/motorSubsystem/MotorSubsystem.java [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+import frc.robot.subsystems.GBSubsystem;
+
+public class MotorSubsystem extends GBSubsystem {
+
+	private final ControllableMotor motor;
+	private final MotorCommandBuilder commandBuilder;
+	private IRequest<Double> voltageRequest;
+	private InputSignal<Double> voltageSignal;
+	private IRequest lastChangedRequest;
+	private boolean wasRequestApllied;
```
Comment (GilStein1) [inline src/main/java/frc/robot/subsystems/motorSubsystem/MotorSubsystem.java:15]: apllied?

File: src/main/java/frc/robot/subsystems/motorSubsystem/MotorSubsystem.java [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+
+	public void updateInputs() {
+		motor.updateInputs(voltageSignal);
+	}
+
+	private void reapplyRequest() {
+		if (!motor.isConnected()) {
+			wasRequestApllied = false;
+		}
+		if (!wasRequestApllied&& motor.isConnected()) {
```
Comment (GilStein1) [inline src/main/java/frc/robot/subsystems/motorSubsystem/MotorSubsystem.java:66]: spotlessApply

File: src/main/java/frc/robot/subsystems/motorSubsystem/MotorCommandBuilder.java [hunk 15]
```diff
@@ -0,0 +1,32 @@
+package frc.robot.subsystems.motorSubsystem;
+
+import edu.wpi.first.wpilibj2.command.Command;
+import edu.wpi.first.wpilibj2.command.RunCommand;
+
+import java.util.function.DoubleSupplier;
+
+public class MotorCommandBuilder {
```
Comment (Nitay4) [inline src/main/java/frc/robot/subsystems/motorSubsystem/MotorCommandBuilder.java:8]:
```suggestion
public class MotorCommandsBuilder {
```

File: src/main/java/frc/robot/subsystems/motorSubsystem/MotorSubsystem.java [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+import frc.utils.alerts.AlertManager;
+import frc.utils.alerts.PeriodicAlert;
+
+public class MotorSubsystem extends GBSubsystem {
+
+	private final ControllableMotor motor;
+	private final MotorCommandBuilder commandBuilder;
+	private IRequest<Double> voltageRequest;
+	private InputSignal<Double> voltageSignal;
+	private IRequest lastChangedRequest;
```
Comment (Nitay4) [inline src/main/java/frc/robot/subsystems/motorSubsystem/MotorSubsystem.java:17]:
```suggestion
	private IRequest<?> lastChangedRequest;
```

File: src/main/java/frc/robot/subsystems/motorSubsystem/MotorSubsystem.java [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+	private boolean wasRequestApllied;
+
+
+	public MotorSubsystem(ControllableMotor motor, String logPath) {
+		super(logPath);
+		this.motor = motor;
+		this.commandBuilder = new MotorCommandBuilder(this);
+		this.wasRequestApllied = true;
```
Comment (Nitay4) [inline src/main/java/frc/robot/subsystems/motorSubsystem/MotorSubsystem.java:25]: i dont see a applyRequest

File: src/main/java/frc/robot/subsystems/motorSubsystem/MotorSubsystem.java [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+		motor.stop();
+	}
+
+	public MotorCommandBuilder getCommandBuilder() {
+		return commandBuilder;
+	}
+
+	public void setVoltage(double voltage) {
+		if (voltageRequest == null) {
+			new Alert(Alert.AlertType.ERROR,"The voltage request is null. try using '.withVoltageControl'").report();
```
Comment (Nitay4) [inline src/main/java/frc/robot/subsystems/motorSubsystem/MotorSubsystem.java:51]: fomat

File: src/main/java/frc/robot/subsystems/motorSubsystem/MotorSubsystem.java [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+		}
+		this.voltageRequest.withSetPoint(voltage);
+		lastChangedRequest = voltageRequest;
```
Comment (Nitay4) [inline src/main/java/frc/robot/subsystems/motorSubsystem/MotorSubsystem.java:54]: else

**Pull request closed without merging**